pyanaconda/modules/boss/module_manager/start_modules.py

```diff
@@ -0,0 +1,145 @@
+#
+# Support for object containers
+#
+# Copyright (C) 2019  Red Hat, Inc.  All rights reserved.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+from queue import SimpleQueue
+
+from pyanaconda.anaconda_loggers import get_module_logger
+from pyanaconda.dbus.constants import DBUS_FLAG_NONE, DBUS_START_REPLY_SUCCESS
+from pyanaconda.dbus.namespace import get_dbus_name
+from pyanaconda.modules.boss.module_manager import ModuleObserver
+from pyanaconda.modules.common.constants.namespaces import ADDONS_NAMESPACE
+from pyanaconda.modules.common.errors.module import UnavailableModuleError
+from pyanaconda.modules.common.task import Task
+
+log = get_module_logger(__name__)
+
+__all__ = ["StartModulesTask"]
+
+
+class StartModulesTask(Task):
+
+    def __init__(self, message_bus, module_names, addons_enabled, timeout=600):
+        super().__init__()
+        self._message_bus = message_bus
+        self._module_names = module_names
+        self._addons_enabled = addons_enabled
+        self._service_timeout = timeout
+        self._module_observers = []
+        self._callbacks = SimpleQueue()
+
+    @property
+    def name(self):
+        """Name of the task."""
+        return "Start the modules"
+
+    def run(self):
+        """Run the task.
+
+        :return: a list of observers
+        """
+        self._module_observers = self._find_modules() + self._find_addons()
+        unavailable = set(self._module_observers)
+        self._start_modules(self._module_observers)
+
+        while unavailable:
+            callback = self._callbacks.get()
+            unavailable.discard(callback())
```

Comment:
I'm not really sure what happens there - I guess a couple comments should be added there to explain it ? Also could this get stuck forever in some cases - I don't see any bounding or timeouts at this stage ?

Reply:
Added comments. There is an explicitly set timeout for the DBus calls and it seems to be enough for now. If a DBus call times out, we get a callback with an error. If a DBus service fails to start, we get a callback with an error. If we receive an error, we raise an exception and quit the task immediately.